src/ascii/chain.rs

```diff
@@ -0,0 +1,386 @@
+        let replies = port
+            .command_replies_until_timeout_with_check("get system.axiscount", check::flag_ok())?;
+        let mut unique_addresses = std::collections::HashSet::with_capacity(replies.len());
+        let mut infos = Vec::with_capacity(replies.len());
+        for reply in replies {
+            // Replies are always from non-zero addresses so it is safe to unwrap.
+            let address: NonZeroU8 = reply.target().device().try_into().unwrap();
+            if !unique_addresses.insert(address) {
+                return Err(DuplicateAddressError {
+                    address: address.get(),
+                }
+                .into());
+            }
+            let axis_count = reply.data().parse().map_err(|_| {
+                AsciiCheckDataError::new(
+                    "could not parse data as 8-bit unsigned integer",
+                    reply.into(),
+                )
+            })?;
+            infos.push(DeviceInfo {
+                address,
+                axis_count,
+            })
+        }
+        // To make future access more efficient sort the list of devices by address.
+        infos.sort_by_key(|info| info.address);
+        Ok(infos)
+    }
+}
+
+/// An iterator of the devices in a [`Chain`].
+#[derive(Debug)]
+pub struct ChainIter<'i, 'a, B, P> {
+    chain: &'i Chain<'a, B, P>,
+    next_device_address: NonZeroU8,
+}
+
+impl<'i, 'a, B, P: Clone> std::iter::Iterator for ChainIter<'i, 'a, B, P> {
+    type Item = Device<'a, B, P>;
+    fn next(&mut self) -> Option<Self::Item> {
+        self.chain
+            .device(self.next_device_address.get())
+            .map(|device| {
+                self.next_device_address = self.next_device_address.saturating_add(1);
+                device
+            })
+    }
+}
+
+/// A by-value iterator over the devices in a chain.
+#[derive(Debug)]
+pub struct ChainIntoIter<'a, B, P> {
+    chain: Chain<'a, B, P>,
+    next_device_address: NonZeroU8,
+}
+
+impl<'a, B, P: Clone> std::iter::Iterator for ChainIntoIter<'a, B, P> {
+    type Item = Device<'a, B, P>;
+    fn next(&mut self) -> Option<Self::Item> {
+        self.chain
+            .device(self.next_device_address.get())
+            .map(|device| {
+                self.next_device_address = self.next_device_address.saturating_add(1);
+                device
+            })
+    }
+}
+
+/// Represents a single device (controller or integrated product).
+#[derive(Debug, Clone)]
+pub struct Device<'a, B, P> {
+    port: P,
+    info: DeviceInfo,
+    _markers: Markers<'a, B>,
+}
+
+impl<'a, B, P> Device<'a, B, P> {
+    /// Get the [`Target`] for the axis.
+    pub fn target(&self) -> Target {
+        self.info.address.get().into()
+    }
+
+    /// Get an iterator over the axes of the device.
+    pub fn iter(&self) -> DeviceIter<'_, 'a, B, P> {
+        DeviceIter {
+            device: self,
+            next_axis_number: unsafe { NonZeroU8::new_unchecked(1) },
+        }
+    }
+}
```

Comment:
Missing methods to send commands. Probably want to mirror the methods on the `Port`. Also applies to `Axis`.

Reply:
May be refactoring some of those commands, so I'm not going to implement them now.